### PR TITLE
Extract judge classifier to ensemble module

### PR DESCRIPTION
**Pull request by gcattan**
This PR moves the JudgeClassifier from the example with brain invader illiteracy to a new module `ensemble.py`.

### CODE REVIEW

File: pyriemann_qiskit/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            Target vector relative to X.
+
+        Returns
+        -------
+        self : JudgeClassifier instance
+            The JudgeClassifier instance.
+        """
+        self.classes_ = np.unique(y)
+        y1 = self.c1.fit(X, y).predict(X)
+        y2 = self.c2.fit(X, y).predict(X)
```
[the remark this answers is not in the document]
Comment (gcattan) [reply pyriemann_qiskit/ensemble.py:71]: Well, this is a good idea. The default behavior could be to take the predictions on which at least one disagree.

File: pyriemann_qiskit/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+from pyriemann_qiskit.utils import union_of_diff
+from sklearn.base import ClassifierMixin, BaseEstimator
+
+
+class JudgeClassifier(BaseEstimator, ClassifierMixin):
+
+    """Judge classifier
+
+    Two (or more) classifiers are trained on the balanced dataset.
+    When the two classifiers disagreed on the label of a given input
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:14]:
```suggestion
    Several classifiers are trained on the balanced dataset.
    When at least one classifier disagrees on the label of a given input
```

File: pyriemann_qiskit/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    in the training set, the input was noted.
+    These inputs, a subset of the training data of the balanced dataset,
+    formed an additional dataset on which a metaclassifier was subsequently
+    trained. Adapted from [1]_.
+
+    Parameters
+    ----------
+    clfs : ClassifierMixin[]
+        An liast of ClassifierMixin.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:23]:
```suggestion
        A list of ClassifierMixin.

```

File: pyriemann_qiskit/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    """
+
+    def __init__(self, judge, clfs):
+        self.clfs = clfs
+        self.judge = judge
+
+    def fit(self, X, y):
+        """Train all classifiers in clfs.
+
+        Then Train the judge classifier on the the samples for which
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:50]:
```suggestion
        Then train the judge classifier on the the samples for which

```

File: pyriemann_qiskit/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        the classifiers have different predictions.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features) | (n_samples, n_features, n_times)
+            The shape of X (vectors or matrices) should be the same for all classifiers
+            (clf1, clf2 and judge).
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:57]:
```suggestion
            (clfs and judge).

```

File: pyriemann_qiskit/ensemble.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    trained. Adapted from [1]_.
+
+    Parameters
+    ----------
+    clfs : ClassifierMixin[]
+        An liast of ClassifierMixin.
+    judge : ClassifierMixin
+        An instance of ClassifierMixin.
+        This classifier is trained on the labels for which
+        clf1 and clf2 obtain different predictions.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:27]:
```suggestion
        classifiers clfs obtain different predictions.

```

File: pyriemann_qiskit/ensemble.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            self.judge.fit(X[mask], y[mask])
+
+    def predict(self, X):
+        """Calculates the predictions.
+
+        When the classifiers don't agree on the prediction
+        the judge classifier is used.
+
+        The behavior is that if at least one of the classifier
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:80]:
```suggestion
        The behavior is that if at least one of the classifiers

```

File: tests/test_ensemble.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    JudgeClassifier,
+)
+from sklearn.svm import SVC
+from sklearn.base import ClassifierMixin
+from sklearn.model_selection import cross_val_score, StratifiedKFold
+import numpy as np
+
+
+def test_canary():
+    assert JudgeClassifier(SVC(), SVC(), SVC()) is not None
```
Comment (qbarthelemy) [inline tests/test_ensemble.py:11]: To fix tests
```suggestion
    assert JudgeClassifier(SVC(), [SVC(), SVC()]) is not None
```

File: tests/test_ensemble.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def test_get_set_params():
+    X = np.array([[0] * 4, [0] * 4, [1] * 4, [1] * 4])
+    y = np.array([0, 0, 1, 1])
+    estimator = JudgeClassifier(SVC(), SVC(), SVC())
```
Comment (qbarthelemy) [inline tests/test_ensemble.py:17]:
```suggestion
    estimator = JudgeClassifier(SVC(), [SVC(), SVC()])
```

**Review by qbarthelemy — COMMENTED** [review]
Great improvement! But tests are broken.

File: pyriemann_qiskit/ensemble.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+class JudgeClassifier(BaseEstimator, ClassifierMixin):
+
+    """Judge classifier
+
+    Several classifiers are trained on the balanced dataset.
+    When at least one classifier disagrees on the label of a given input
+    in the training set, the input was noted.
+    These inputs, a subset of the training data of the balanced dataset,
+    formed an additional dataset on which a metaclassifier was subsequently
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:17]: What is the difference between classifier and metaclassifier?
```suggestion
    formed an additional dataset on which a new classifier was subsequently

```

File: pyriemann_qiskit/ensemble.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(self, judge, clfs):
+        self.clfs = clfs
+        self.judge = judge
+
+    def fit(self, X, y):
+        """Train all classifiers in clfs.
+
+        Then train the judge classifier on the the samples for which
+
+        the classifiers have different predictions.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:54]:
```suggestion
        Then train the judge classifier on the the samples for which
        the classifiers have different predictions.

```

File: pyriemann_qiskit/ensemble.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    def predict_proba(self, X):
+        """Return the probabilities associated with predictions.
+
+        When classifiers clfs have the same prediction, the
+        returned probability is the average of the probability of classifiers.
+
+
+        When classifiers don't have the same predictions,
+
+        the returned probability is the the one of the judge classifier.
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:122]:
```suggestion
        When classifiers don't have the same predictions,
        the returned probability is the the one of the judge classifier.

```

**Review by qbarthelemy — COMMENTED** [review]
Can you complete tests with a case with three classifiers in clfs?

File: pyriemann_qiskit/ensemble.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+class JudgeClassifier(BaseEstimator, ClassifierMixin):
+
+    """Judge classifier
+
+    Several classifiers are trained on the balanced dataset.
+    When at least one classifier disagrees on the label of a given input
+    in the training set, the input was noted.
+    These inputs, a subset of the training data of the balanced dataset,
+    formed an additional dataset on which a new classifier was subsequently
+
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:18]:
```suggestion
    formed an additional dataset on which a new classifier was subsequently

```

File: pyriemann_qiskit/ensemble.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    def predict(self, X):
+        """Calculates the predictions.
+
+        When the classifiers don't agree on the prediction
+        the judge classifier is used.
+
+        The behavior is that if at least one of the classifiers
+
+        doesn't have the same prediction for a particular sample,
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:87]:
```suggestion
        The behavior is that if at least one of the classifiers
        doesn't have the same prediction for a particular sample,
```

File: pyriemann_qiskit/ensemble.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        y_pred[mask] = self.judge.predict(X_diff)
+        return y_pred
+
+    def predict_proba(self, X):
+        """Return the probabilities associated with predictions.
+
+        When classifiers clfs have the same prediction, the
+        returned probability is the average of the probability of classifiers.
+
+
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:120]:
```suggestion
```

File: pyriemann_qiskit/ensemble.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    ----------
+    clfs : ClassifierMixin[]
+        A list of ClassifierMixin.
+
+    judge : ClassifierMixin
+        An instance of ClassifierMixin.
+        This classifier is trained on the labels for which
+        classifiers clfs obtain different predictions.
+
+
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:31]:
```suggestion
```

File: pyriemann_qiskit/ensemble.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        """Return the probabilities associated with predictions.
+
+        When classifiers clfs have the same prediction, the
+        returned probability is the average of the probability of classifiers.
+
+        When classifiers don't have the same predictions,
+        the returned probability is the the one of the judge classifier.
+
+
+        Parameters
```
Comment (qbarthelemy) [inline pyriemann_qiskit/ensemble.py:121]:
```suggestion
        returned probability is the average of the probability of classifiers.
        When classifiers don't have the same predictions,
        the returned probability is the the one of the judge classifier.

        Parameters
```

**Review by qbarthelemy — APPROVED** [review]
Great job @gcattan !

**Comment by gcattan** [comment]
Thank you for the review @qbarthelemy :)

We are working on presenting some results on Q-MDM with @toncho11 
Is it ok to send you a mail on your GH email?

**Comment by qbarthelemy** [comment]
> Is it ok to send you a mail on your GH email?
Yes, it's ok. @gcattan

**Pull request merged**